docs/learn/test-runner/writing-plugins.md

```diff
@@ -6,9 +6,301 @@ eleventyNavigation:
+  window.process = { env: { NODE_ENV: 'development' } };
+</script>
+```
+
+In order to insert this snippet, we can us a transform plugin and insert it right into the header.
+
+```js
+export default {
+  files: 'test/**/*.test.js',
+  nodeResolve: true,
+  plugins: [
+    {
+      name: 'provide-process',
+      transform(context) {
+        if (context.path === '/') {
+          const transformedBody = context.body.replace(
+            '</head>',
+            '<script>window.process = { env: { NODE_ENV: "development" } }</script></head>',
+          );
+          return transformedBody;
+        }
+      },
+    },
+  ],
+};
+```
+
+With that, we are finally green 💪
+
+```
+Chrome: |██████████████████████████████| 1/1 test files | 1 passed, 0 failed
+
+Finished running tests in 1.3s, all tests passed! 🎉
+```
+
+Often when using 3rd party npm dependencies the code might not work out of the box in a browser.
+With this plugin we however made it possible to run popper directly in the browser and therefore also within our test runner.
+
+## Use case mocking API
+
+Next up we are having a helper function that can get meta information from users.
+This function requires an api as it will fetch a REST json endpoint while in production.
+
+The test is checking if the data you get is correct.
+
+👉 `test/getUser.test.js`
+
+```js
+it('can get meta data via a userId', async () => {
+  const user3 = await getUser(3);
+  expect(user3.name).to.qual('Lea');
+
+  const user10 = await getUser(10);
+  expect(user10.name).to.qual('Peter');
+});
+```
+
+Now when we implement we use fetch to actually toggle the api endpoint.
+
+👉 `src/getUser.js`
+
+```js
+export async function getUser(userId) {
+  const response = await fetch(`/api/users/${userId}`);
+  const data = await response.json();
+  return data;
+}
+```
+
+When we start the test we do get an expected error.
+
+```
+test/getUser.test.js:
+
+ 🚧 404 network requests:
+    - api/users/3
+
+ ❌ can get meta data via a userId
+      SyntaxError: Unexpected token N in JSON at position 0
+```
+
+The url `api/users/3` returned `404` which means the file could not be found.
+Also, our json parsing failed as the `404` page has `Not Found` as its body.
+And `Not Found` is not valid json 😅
+
+To create virtual files we can use the `serve` plugin hook.
+In there again you have access to the `koa context`.
+
+> By default the file extension is used to infer the mime type to respond with.
+
+As we are using no file extension at all we need to return an object with a `type` property to set it explicitly.
+
+```js
+export default {
+  files: 'test/**/*.test.js',
+  nodeResolve: true,
+  plugins: [
+    {
+      name: 'mock-api',
+      serve(context) {
+        if (context.path === '/api/users/3') {
+          return { body: '{ "name": "Lea" }', type: 'json' };
+        }
+        if (context.path === '/api/users/10') {
+          return { body: '{ "name": "Peter" }', type: 'json' };
+        }
+      },
+    },
+  ],
+};
```

Comment:
This is not actually what a plugin is for, the serve hook is for serving files. You're actually responding with an API request. That's what middleware are for.

Reply:
But it's also not wrong, and maybe it's a nice convenience way to do it.

Reply:
hmmm if that is not how we should do it then we need to make clear why? (I didn't get it apparently 🙈)

what is the difference to a middelware? when should you use a middelware? when a plugin?

right now we don't even teach about middleware at all 😅 
I thought we encourage to always use plugins 🤔

Reply:
added a note